README
GNU Source Release Collection (GSRC)
####################################

Introduction
------------

The GNU Source Release Collection provides a simple way to install the
latest GNU packages (and more) in their own directory (on an existing system).

This is particularly useful when you want to install them on
a system for which you do not have administrative rights, such as a
shared cluster.  It simplifies the process of maintaining a collection
of locally installed software by cleanly separating the files of each
package (and each package version), and installing symbolic links to
the files in the directory of your choice.

Bioinformatics Source Release Collection is
now part of GNU Source Release Collection GSRC.

### Getting started

First, get it. 

bzr checkout bzr://bzr.savannah.gnu.org/gsrc/trunk/ gsrc

( or download a tar file from ftp://ftp.gnu.org/gnu, but
bzr method will bring up-to-date files, which change daily.)
 

To get started, you must first configure it. From a shell, type:

    $ ./bootstrap
    $ ./configure --prefix=$HOME/gnu

You may choose to install the GNU software to any location you like by
setting the `--prefix' argument (they will be installed to ~/gnu in
the above example).

To update, in the gsrc directory, just do:

bzr update

You will find the available GNU packages under the `gnu'
sub-directory. Since they encompass a large number of packages in
their own right, the GNUstep and GNOME projects have their own
sub-directories: `gnustep' and `gnome', respectively.
Also 'bio' for bioinformatics, and 'other' for whatever else.

To get help on getting started, type "make help".

Status
------

For a complete listing of packages and their current status in GSRC,
see https://www.gnu.org/software/gsrc/package-list.html


Bugs and build failures
-----------------------

If you encounter a package that does not build properly, please first
check its status in `TODO'; there are a few packages which are known to
not build properly. If the affected package is listed as `DONE' and
not `BROKEN', then this should be reported. Similarly, if you find any
outdated packages, please report them. All problems can be sent to the
GSRC bug mailing list <bug-gsrc@gnu.org>.

## Featured software

If a program is missing from GSRC or BioSRC, please send a message to the
mailing list bug-gsrc@gnu.org  to request its addition.
Similarly, if you notice that a program version is out-of-date, please
send a message.

Only software that is licensed under an accepted free software license
will be included in GSRC/BioSRC.  Lamentably, many bioinformatics programs
are released in binary format only, under proprietary licenses, under
otherwise restrictive licenses ("...for academic use only...") or,
just as bad, with no explicit license at all (which effectively makes
the software proprietary).  Distributing binary-only files is
tantamount to hiding your scientific methods, while restrictive,
proprietary licenses undermine the principles of open and
collaborative science.  Thus, the use of such programs should not be
encouraged.  Finally, anyone installing software with BioSRC should be
confident that they have full permission to use the software in any
way they like.  We don't want to provide software that not everyone
can use!  If you wish to have a program added to BioSRC that is
proprietary or under a restrictive license, please contact the
software's author(s) to encourage them to re-release the software
under an appropriate license.

## Supported systems

GSCR/BioSRC is geared towards installing software on GNU/Linux systems.
It may be possible to use it on other Unix-oid operating systems,
however note that you might need to modify some files to make them work.
Patches to support building on other systems will only be accepted if
they do not disrupt building on GNU/Linux systems.

Windows?  You might have some luck with CygWin, but no support can be
given in that regard.



Copying
-------

GSRC is based on the GAR build system by Nick Moffitt and GARstow
enhancements by Adam Sampson.

Files from GAR and GARstow are under the following license unless
otherwise noted:

  Copyright (C) 2001 Nick Moffitt
  Copyright (C) 2004, 2005, 2006, 2007, 2008, 2009, 2010 Adam Sampson

  Redistribution and/or use, with or without modification, is
  permitted.  This software is without warranty of any kind.  The
  author(s) shall not be liable in the event that use of the
  software causes damage.

Files specific to GSRC, including the build recipes, are Copyright (C)
Free Software Foundation and licensed under the GNU General Public
License, Version 3, or (at your option) any later version. Please see
the file `COPYING' for more information.
 

<!-- Copyright (C) 2014 Brandon Invergo <brandon@invergo.net> -->

<!-- Author: Brandon Invergo <brandon@invergo.net> -->

<!-- This program is free software; you can redistribute it and/or -->
<!-- modify it under the terms of the GNU General Public License -->
<!-- as published by the Free Software Foundation; either version 3 -->
<!-- of the License, or (at your option) any later version. -->

<!-- This program is distributed in the hope that it will be useful, -->
<!-- but WITHOUT ANY WARRANTY; without even the implied warranty of -->
<!-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the -->
<!-- GNU General Public License for more details. -->

<!-- You should have received a copy of the GNU General Public License -->
<!-- along with this program. If not, see <http://www.gnu.org/licenses/>. -->


## Copying

BioSRC is based on the GNU Source Release Collection (GSRC), which is
in turn based on the GAR build system by Nick Moffitt and GARstow
enhancements by Adam Sampson.

Files specific to BioSRC, including the build recipes, are Copyright
(C) Brandon Invergo and licensed under the GNU General Public License,
Version 3, or (at your option) any later version. Please see the file
`COPYING` for more information.
 


NEWS
# -*- org -*-

Copyright (C) 2012, 2013, 2014 Brandon Invergo <brandon@invergo.net>

This file is part of GSRC.

GSRC is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

GSRC is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with GSRC.  If not, see <http://www.gnu.org/licenses/>.

# ****************************************************************** #
# GSRC is best used as a "rolling-release" system, in which the user #
# stays up-to-date with the latest changes via a personal bzr        #
# repository. Alternatively, every quarter, a snapshot is released,  #
# which may be downloaded and used on its own.                       #
#                                                                    #
# This file outlines the updates that have occurred between snapshot #
# releases.                                                          #
#                                                                    #
# Note: this file is best viewed in Emacs Org-Mode                   #
# ****************************************************************** #

* GSRC 2016.06.17 
** Changes to the GSRC system
 file layout changed. BioSRC merged.
 see "IMPORTANT.MESSAGE" file

* GSRC 2014.10.11

** Changes to the GSRC system

*** Bugfix: Uninstallation of packages without info files

GSRC will no longer try to remove info files if they aren't
installed.  This would prevent the uninstallation procedure from
completing.

*** Bugfix: Trailing whitespace from relevant package metadata is stripped

Trailing whitespace on the GARNAME, GARPROFILE, and GARVERSION
variables is explicitly stripped to prevent bugs when the shell tries
to operate on path names derived from some combination of their
values.

** Changes in GSRC packages

*** 2 packages have been added to GSRC since the last release

*** 49 packages have been updated since the last release

*** Coverage statistics
Note: the totals include packages that will not be implemented in
GSRC, such as retired packages, which are counted as completed. Broken
builds are counted as incomplete. Some packages have not yet made any
releases, which are also counted as incomplete. See TODO for more
information. 

**** GNU packages [424/473] [89%]
**** GNOME packages
***** Core [15/116] [12%]
***** Apps [1/44] [2%]
**** GNUstep packages [22/28] [78%]

*** New packages
(most recent version given)

**** datamash (1.0.5)
**** direvent (5.0)

*** Updated packages
(most recent version given)

**** apl (1.4)
**** bash (4.3-30)
**** chess (6.1.2)
**** coreutils (8.23)
**** cssc (1.4.0)
**** datamash (1.0.6)
**** ddrescue (1.19)
**** denemo (1.1.8)
**** easejs (0.2.4)
**** emms (4.0)
**** freeipmi (1.4.5)
**** gcc (4.9.1)
**** gcl (2.6.11)
**** gdb (7.8)
**** gettext (0.19.2)
**** gforth (0.7.3)
**** global (6.3.2)
**** glpk (4.55)
**** gnubg (1.03.001)
**** gnucash (2.6.4)
**** gnudos (1.7)
**** gnupg (2.0.26)
**** gnutls (3.3.8)
**** gss (1.0.3)
**** gtypist (2.9.5)
**** guile-ncurses (1.5)
**** help2man (1.46.3)
**** libc (2.20)
**** libgcrypt (1.6.2)
**** libidn (1.29)
**** librejs (6.0.1)
**** libtasn1 (4.2)
**** libunistring (0.9.4)
**** lightning (2.0.5)
**** lilypond (2.19.1)
**** linux-libre (3.16)
**** make (4.1)
**** mdk (1.2.8)
**** mediagoblin (0.7.1)
**** nano (2.3.6)
**** ocrad (0.24)
**** octave (3.8.2)
**** parallel (20140922)
**** parted (3.2)
**** pspp (0.8.4)
**** rcs (5.9.3)
**** swbis (1.12)
**** tar (1.28)
**** ucommon (6.1.11)

* GSRC 2014.07.06

** Changes to the GSRC system

*** Using GARPROFILE correctly modifies the package .gar directory

** Changes in GSRC packages

*** 4 packages have been added to GSRC since the last release

*** 61 packages have been updated since the last release

*** Coverage statistics
Note: the totals include packages that will not be implemented in
GSRC, such as retired packages, which are counted as completed. Broken
builds are counted as incomplete. Some packages have not yet made any
releases, which are also counted as incomplete. See TODO for more
information. 

**** GNU packages [422/471] [89%]
**** GNOME packages
***** Core [15/116] [12%]
***** Apps [1/44] [2%]
**** GNUstep packages [22/28] [78%]

*** New packages
(most recent version given)

**** easejs (0.2.1)
**** gnudos (1.0)
**** guile-opengl (0.1.0)
**** guile-rpc (0.4)

*** Updated packages
(most recent version given)

**** anubis (4.2)
**** autogen (5.18.3)
**** bash (4.3-18)
**** ccaudio (2.1.3)
**** ccrtp (2.0.9)
**** ccscript (5.0.2)
**** cgicc (3.2.15)
**** cim (5.1)
**** cursynth (1.5)
**** dap (3.10)
**** ddrescue (1.18.1)
**** denemo (1.1.6)
**** easejs (0.2.2)
**** electric (9.05)
**** freedink (108.2)
**** freeipmi (1.4.4)
**** fribidi (0.19.6)
**** gawk (4.1.1)
**** gcc (4.9.0)
**** gcompris (14.05)
**** gdb (7.7.1)
**** gettext (0.19.1)
**** global (6.3)
**** gnubatch (1.11)
**** gnun (0.9)
**** gnunet (0.10.1)
**** gnunet-gtk (0.10.1)
**** gnupg (2.0.25)
**** gnustandards (2014)
**** gnutls (3.3.4)
**** grep (2.20)
**** gtick (0.5.3)
**** guile-dbi (2.1.5)
**** libmicrohttpd (0.9.37)
**** libtasn1 (4.0)
**** lightning (2.0.4)
**** lilypond (2.19.1)
**** linux-libre (3.15)
**** mailman (2.1.18)
**** mcron (1.0.8)
**** mediagoblin (0.6.1)
**** melting (5.1.2)
**** mifluz (0.26.0)
**** mit-scheme (9.2)
**** moe (1.6)
**** nano (2.3.4)
**** nettle (3.0)
**** parallel (20140622)
**** pspp (0.8.3)
**** pyconfigure (0.2.2)
**** r (3.1.0)
**** screen (4.2.1)
**** sipwitch (1.9.1)
**** sqltutor (0.7)
**** texmacs (1.0.7.21)
**** ucommon (6.1.5)
**** units (2.11)
**** wdiff (1.2.2)
**** xaos (3.6)
**** xnee (3.19)
**** xorriso (1.3.8)


util/fetchupdate.pl
#!/usr/bin/perl
$ENV{'FTP_PASSIVE'}=1;
$debug=1;
use Term::ReadLine;
use Data::Dumper;
use LWP::Simple;
use LWP::UserAgent;

#print versioncmp("mailman-2.1.13rc1","mailman-2.1.13");
#exit;

$\=undef;
$,=" ";

$dir = $ARGV[0];
$dir =~ s/\/$//;

$timestamp = "$dir/.update-timestamp";
if (-e $timestamp && -M $timestamp < 0.1) {
    print "Checked recently\n";
    exit (0);
};

$r_generic = get_make_info($dir,"GARVERSION=\@VERSION\@","GARNAME=\@NAME\@");

print Dumper($r_generic);

$r = get_make_info($dir);

$name = $r->{'Name'};
$orig_version = $r->{'Version'};
$orig_file = $r->{'Distribution files'}->[0];
$orig_url = $r->{'Location'} . $orig_file;
print "Original url is $orig_url\n";
my $orig_res = head($orig_url); 

$generic_directory = $r_generic->{'Location'} ;
if ($generic_directory =~ /\@VERSION\@/) {
    $directory = search_generic($generic_directory);
    print "Directory = $directory\n";
} else {
    $directory = $r->{'Location'} ;
}
 
print "Checking directory $directory\n";
my $res = get($directory); 
($new_file, $sig) =  find_latest_file($res); 

if ($new_file eq "") {
    print $res;
    die "can't find files";
};

$directory =~ s/\/$//;
$new_url = $directory."/".$new_file;

if ($new_url eq $orig_url) {
    print "File is up to date\n";
    open(TIMESTAMP,">$timestamp"); close(TIMESTAMP);
    exit;
}

print "Comparing old and new:\n";
print "orig url = <$orig_url>\n";
print "new url = <$new_url>\n";
$new_res = head($new_url); 
print Dumper($new_res);
print $orig_file, ": ", $orig_res->header('last-modified'), "\n";
print $new_file, ": ", $new_res->header('last-modified'),  "\n";

$generic_name = $r_generic->{'Distribution files'}->[0];
$version_capture = $generic_name;
$version_capture =~ s/\./\\./g;
$version_capture =~ s/(\@NAME\@)/$name/g;
$version_capture =~ s/(\@VERSION\@)/(.*)/g;
print "regex = $version_capture\n";
($new_version) = $new_file =~ /$version_capture/;
print "GARVERSION = $orig_version -> $new_version\n";
if ($new_version eq "") {
    die "can't extract new version";
};

update_garversion ($dir, $orig_version, $new_version);

sub update_garversion {
    my $dir = shift;
    my $orig_version = shift;
    my $new_version = shift;
    my $makefile = "$dir/Makefile";
    open (INPUT, "<$makefile") || die "cannot open $makefile: $!";
    my @lines = <INPUT>;
    close(INPUT);
    my @matches = grep(s/^(\s*(?:override\s+)?GARVERSION\s*\??=\s*)$orig_version(\s*)/$1$new_version$2/,@lines);
    if (@matches == 0) {
        die "no matches for $orig_version in $makefile";
    } elsif (@matches > 1) {
        die "multiple matches for $orig_version in $makefile", @matches;
    }

    open(OUTPUT,">$makefile.new") || die "cannot open $makefile.new: $!";
    print OUTPUT join("", @lines);
    close(OUTPUT);

    $r_new = get_make_info($dir,"-f Makefile.new");
    
    mysystem("diff -u $makefile $makefile.new");
    mysystem("cp  $makefile.new .update-pending");
    mysystem("mv -b $makefile.new $makefile");

    open(OUTPUT,">$dir/logmsg") || die "cannot open $dir/logmsg: $!";
    print OUTPUT "updated $name from $orig_version to $new_version\n";
    close(OUTPUT);
}

sub search_generic {
    print "Generic directory is $generic_directory\n";
    $directory = $generic_directory;
    my @dirs = split('/', $generic_directory);
    do {
        $subdir = pop @dirs;
        $directory = join("/", @dirs);
        print " -> $directory\n";
    } while ($directory =~ /\@VERSION\@/);
    print "Base directory $directory\n";
    print "Subdir pattern $subdir\n";
    $subdir =~ s/\@VERSION\@/.*/g;
    print "-> $subdir\n";
    print "Checking $directory\n";
    my $res = get($directory); 
    return $directory . "/" . find_latest_dir($res,$subdir);
}

sub get_make_info {
    my $dir = shift @_;
    my @args = @_;
    open(MAKE_OUTPUT, "make -C $dir fetch-list " . join(" ", @args) . "|") || die "can't open $file: $!";
    my $result = parse(<MAKE_OUTPUT>);
    close(MAKE_OUTPUT);
    return $result;
}

sub parse {
    my @lines = @_;
    my $key; 
    my $value;
    my %result;
    for (@lines) {
        next if /^make(\[\d+\])?:/;
        if (/^([ \w]+):\s*$/)  {
            $key = $1 ;
            $value = [];
            $result{$key} = $value;
        } elsif (/^\s+(\S+)/) {
            $value = $1;
            push(@{$result{$key}}, $value);
        } elsif (/^([ \w]+):\s*(\S+)$/)  {
            $key = $1;
            $value = $2;
            $result{$key} .= $value;
        } else {
            die "unrecognized line $_";
        }
    }
    return \%result;
}

######################################################################

sub find_latest_dir {
    my ($output,$subdir) = @_;
    #print "$output\n";
    my @dirs = $output =~ /^d.* (\S*)$/mg;
    print "dirs = ", @dirs, "\n";

    my @candidates = grep(/$subdir/i                        
                          && !/latest/, @dirs);

    debug("candidates = ", @candidates, "\n");
    @candidates = versionsort(@candidates);
    my $latest = @candidates[-1]; 
    print "latest = $latest\n";
    return $latest;
}

sub find_latest_file {
    my ($output) = @_;
    debug("$output\n");
    my @files = $output =~ /([\w\-\.]+\.(?:tar\.[\w\.]+|tgz(?:\.[\w\-\.]+)?))/mg;
    debug("files = ", map("<$_>", @files), "\n");

    my @candidates = grep(/^$name-\d[\w\.]+$/i 
                          && !/\.(diff|patch|xdelta|html)\./
                          #&& !/\.\d+rc\d+\./
                          && !/\d(alpha|beta)/
                          && !/latest/, @files);

    debug("candidates = ", map("<$_>", @candidates), "\n");
    my $latest = (getversions(@candidates))[-1]; 
    print "latest = $latest\n";

    my @selected_files = grep(/^$latest/, @candidates);
    print "selected = ", @selected_files, "\n";

    my @distfiles = grep(!/\.(sig|asc)$/, @selected_files);
    my @sigfiles = grep(/\.(sig|asc)$/, @selected_files);
    print "distfiles = ", @distfiles, "\n";
    print "sigfiles = ", @sigfiles, "\n";

    my %seen = ();
    map($seen{$_} =1 , @distfiles);
    map($seen{$_} =1 , @sigfiles);

    my $distfile = $distfiles[0] ;
    my $sigfile = $sigfiles[0] ;

    for my $ext ('.tar.gz', '.tar.bz2', '.tar.Z', '.tar.xz', '.tar.lzma', '.tgz') {
        if ($seen{"$latest$ext"}) {
            for my $sigext('.sig', '.asc') {
                $distfile = "$latest$ext";
                if ($seen{"$distfile$sigext"}) {
                    $sigfile = "$distfile$sigext";
                    goto DONE;
                };
            }
        }
    }

  DONE:

    print "Latest <$distfile> <$sigfile>\n";
    return ($distfile, $sigfile);
}

sub getversions {
    my @files = @_;
    my %seen;
    for (grep(s/(\.tar\.(gz|Z|bz2|lzma|xz)|\.tgz)$//, @files)) { $seen{$_}=1; } ;
    return versionsort(keys %seen);
}

sub versionsort {
    my @files = @_;
    #print "sorting:\n", join("\n", @files), "\n";
    my @result =  sort { versioncmp($a,$b) } @files;
    #print "result:\n", join("\n", @result), "\n";
    return @result;
}

# $Id: Versions.pm,v 1.9 2003/08/24 22:58:14 ed Exp $

# Copyright (c) 1996, Kenneth J. Albanowski. All rights reserved.  This
# program is free software; you can redistribute it and/or modify it under
# the same terms as Perl itself.

sub versioncmp( $$ ) {
    map(s/^rel//, @_) ; # handle autogen versions rel5.6.7

    my @A = ($_[0] =~ /([-\.]|\d+|[^-\.\d]+)/g);
    my @B = ($_[1] =~ /([-\.]|\d+|[^-\.\d]+)/g);

    my ($A, $B);

    while (@A and @B) {
        $A = shift @A;
        $B = shift @B;
        #print "COMPARE <$A> <$B>\n";
        if ($A eq '-' and $B eq '-') {
            next;
        } elsif ( $A eq '-' ) {
            return -1;
        } elsif ( $B eq '-') {
            return 1;
        } elsif ($A eq '.' and $B eq '.') {
            next;
        } elsif ($A eq 'pre' and $B eq 'pre') {
            next;
        } elsif ( $A eq 'pre') {
            return -1;
        } elsif ( $B eq 'pre') {
            return 1;
        } elsif ($A eq 'rc' and $B eq 'rc') {
            next;
        } elsif ( $A eq 'rc') {
            return -1;
        } elsif ( $B eq 'rc') {
            return 1;
        } elsif ( $A eq '.') {
            return -1;
        } elsif ( $B eq '.') {
            return 1;
        } elsif ($A =~ /^\d+$/ and $B =~ /^\d+$/) {
            if ($A =~ /^0/ || $B =~ /^0/) {
                return $A cmp $B if $A cmp $B;
            } else {
                return $A <=> $B if $A <=> $B;
            }
        } else {
            $A = uc $A; 
            $B = uc $B; 
            return $A cmp $B if $A cmp $B;
        }
    }

    if (@A && $A[0] =~ /pre|rc/i) {
        return -1;
    } 

    if (@B && $B[0] =~ /pre|rc/i) {
        return +1;
    } 

    @A <=> @B;
}

sub debug {
    print @_ if $debug;
}

sub mysystem {
    my @lines;
    print "+ ", join(" ", @_), "\n";
    open(CMD,"-|", join(" ", @_) . " 2>&1" ) || die;
    while(<CMD>) { print "$_"; push (@lines, $_); } ;
    close(CMD);
    #system(@_) ;

    if ($? == -1) {
	print "failed to execute: $!\n";
    }
    elsif ($? & 127) {
	printf "child died with signal %d, %s coredump\n",
	($? & 127),  ($? & 128) ? 'with' : 'without';
    }
    elsif ($? >> 8) {
	printf "child exited with value %d\n", $? >> 8;
    }
    
    sleep(1);
    return ($?, @lines);
}


util/buildit.sh
#!/usr/bin/perl
$|=1;

use Getopt::Long;
my $PREFIX;
my $opt_remove_all;
$result = GetOptions ("prefix=s" => \$PREFIX,
                      "remove-all" => \$opt_remove_all);

die "need prefix directory --prefix" if ! -d $PREFIX;

$LOGDIR="logs"; mkdir($LOGDIR) if ! -d $LOGDIR;

open(SECRET,"<$ENV{HOME}/.secret") || die; $secret = <SECRET>; chomp($secret); close(SECRET);

@pkgs = @ARGV ;
@pkgs = grep(-d, glob("gnu/*")) if !@pkgs;
##@pkgs = random_sort (@pkgs);

for my $pkg (@pkgs) { 
    doit ($pkg);
};

sub doit {
    my $name = shift; $name =~ s/\/$//;
    my $pkg = find_pkg($name) ;

    print "directory is $pkg\n";

    my $deps=`make -C $pkg --no-print-directory  dep-list`;
    my @deps = split(' ', $deps);
    print "$name deps: ", join(" ", @deps), "\n";

    my @failed_deps = ();
    for my $dep (@deps) {
	my $status = doit($dep);
	if ($status) {
	    push (@failed_deps, $dep);
	}
    }
    
    print "$name failed deps = ", join(" ", @failed_deps), "\n";

    my $hash = compute_hash($name, @deps);
    my $logfile = logfile($pkg);
    my $hashfile = hashfile($pkg);
    my $stored_hash = read_hash($hashfile);

    print "new build hash = $hash\n";
    print "old build hash = $stored_hash\n";

    if ($hash eq $stored_hash) {
	if (-f $logfile) {
	    print "$name already run for $pkg\n";
	    return 0;
	} elsif (-f "$logfile.FAIL" && $hash eq $stored_hash) {
	    print "$name already failed for $pkg\n";
	    return -1;
	}
    }

    &reset_installed if $opt_remove_all ;
    mysystem("make -C $pkg uninstall-pkg > /dev/null 2>&1");
    mysystem("make -C $pkg clean > /dev/null 2>&1");

    my $tmp_logfile = "$logfile.$$";
    open(LOG, ">$tmp_logfile") or die;

    my $tmp_hashfile = "$hashfile.$$";
    open(HASH, ">$tmp_hashfile") or die;
    print HASH $hash, "\n";
    close(HASH);

    my @time = gmtime();
    printf LOG "# Build %s %04d-%02d-%02d-%02d%02d\n", $pkg, $time[5]+1900,$time[4]+1,$time[3],$time[2],$time[1];

    my $status; my @lines; my $finalstage; my $autobuild; 
    my $disable_tests = 0;
    my $total_time = 0;
    my $shortname = $pkg; $shortname =~ s/.*\///;
    my @GSRCLOG;
    my @successes; 
    my @failures;
    my $installed;

    for my $stage ("fetch-list", "deps", "fetch", "checksum", "extract", "configure", "build", "test", "install") {
	my $t0 = time;
	if ($stage eq "deps") {
	    $status = scalar(@failed_deps);
	    @lines = map("failed dependency on $_\n", @failed_deps);
	    print @lines;
	    if (@failed_deps) {
		push(@GSRCLOG, "GSRC: $shortname failed-deps " . join(" ", @failed_deps) . "\n");
	    };
	} else {
	    if ($stage =~ /build|test/) {
		$timeout = 15 * 60;
	    } elsif ($stage =~ /configure/) {
		$timeout = 2*60;
	    };

	    my $make = ($stage =~ /build|test/) ? "make -k" : "make";

	    ($status, @lines) = mysystem("$make -C $pkg $stage" . ($disable_tests ? " USE_TESTS=" : ""));
	};
	my $t1 = time;
	$total_time += $t1 - $t0;
	print LOG @lines ;

	$autobuild = 1 if grep(/autobuild project/,@lines);
	$configure_opts = &extract_options(@lines) if $stage eq "configure";

	push(@GSRCLOG, sprintf "GSRC: $shortname stage=$stage time=%ds status=%s (%d)\n", $t1-$t0, ($status ? "FAILED" : "OK"), $status);

	if ($status == 0) {
	    push(@successes, $stage) ;
	} else {
	    push(@failures, $stage) ;
	}

	if ($status && $stage eq "test") { 
	    $disable_tests = 1 ; 
	    ($status, @lines) = mysystem("make -C $pkg test-logs");
	    print LOG @lines;
	    next; 
	} ;

	$installed = 1 if $stage eq "install" && $status == 0;

	last if $status;
    };
    
    if (!$installed) {
	$result = "FAILED/" . join(",", @failures);
    } else {
	$result = "OK/". $successes[-1];
    };

    push(@GSRCLOG, sprintf "GSRC: $shortname result=%s total=%ds\n", $result, $total_time);
    print LOG "\n";
    print LOG @GSRCLOG;
    close(LOG);

    unlink($logfile, "$logfile.FAIL");
    $logfile .= ".FAIL" if !$installed;
    rename($tmp_logfile, $logfile);
    rename($tmp_hashfile, $hashfile);

    mysystem("make -C $pkg clean > /dev/null 2>&1");
    mysystem("gzip -c $logfile | gpg -a -s -e -r bjg\@chapters.gnu.org | Mail -v -s \"$pkg Gsrc log output $secret\" bjg\@chapters.gnu.org");
    if ($autobuild) {
	mysystem("Mail -v $shortname\@autobuild.josefsson.org < $logfile");
    } else {
	my $host = `hostname`; chomp($host);
	my $date = `date '+%Y%m%d-%H%M%S'`; chomp($date);
	my $system = `uname -m`;
	my $mode = $configure_opts // "default";
#	open(MAIL, "|-", "Mail -v $shortname\@autobuild.josefsson.org");
	#open(MAIL, "|-", "Mail -v bjg\@network-theory.co.uk");
	print MAIL "autobuild project... $shortname\n";
	print MAIL "autobuild hostname... $host\n";
	print MAIL "autobuild timestamp... $date\n";
	print MAIL "autobuild mode... $mode\n\n";
	open(LOGFILE,"<$logfile") ; while (<LOGFILE>) { print MAIL $_; } ; close(LOGFILE);
	close(MAIL);
    };
    print "status = $status\n";
    return $status;
}

sub logfile {
    my $pkg = shift;  $pkg =~ s/.*\///;
    return "$LOGDIR/$pkg.log";
}

sub hashfile {
    my $pkg = shift;  $pkg =~ s/.*\///;
    return "$LOGDIR/$pkg.hash";
}


sub mysystem {
    my @lines;
    my $command =  "+ " .  join(" ", @_) .  "\n";
    print $command;
    push (@lines, $command);
    my $pid = open(CMD,"-|", join(" ", @_) . " 2>&1" ) || die;
    print "subprocess pid = $pid\n";
    my $timeout;

    if ($main::timeout) {
	$timeout = $main::timeout;
	$main::timeout = undef;
    } else {
	$timeout = 5*60; # 5 minutes
    };

    eval {
	local $SIG{ALRM} = sub { die "alarm\n" };
	alarm $timeout;
	while(<CMD>) { print "$_"; push (@lines, $_); alarm $timeout} ;
	close(CMD);
	alarm 0;
    };

    if ($@ eq "alarm\n") {
	print "process $pid timed out\n";
	print `ps -o pid,pgrp,user,args`;
	sleep (1);
	my @children = find_children_recursive($pid);
	my @detached = find_children_recursive(1);
	for my $childpid (@children, @detached) { 
	    print "killing process $childpid\n";
	    kill 2, $childpid;
	    sleep (1);
	    kill 9, $childpid;
	};
	print `ps -o pid,pgrp,user,args`;
	push(@lines, "ERROR: Terminated -- process killed by build, no response after $timeout seconds\n");
	return (-1, @lines);
    }

    #system(@_);

    if ($? == -1) {
        print "failed to execute: $!\n";
    }
    elsif ($? & 127) {
        printf "child died with signal %d, %s coredump\n",
        ($? & 127),  ($? & 128) ? 'with' : 'without';
    }
    elsif ($? >> 8) {
        printf "child exited with value %d\n", $? >> 8;
    }

    sleep(1);
    return ($?, @lines);
}

{
    my %lookup = () ;
    sub find_package {
	my $name = shift;
	return $lookup{$name} if defined($lookup{$name});
	for my $p ("$name", glob("*/$name")) {
	    if (-d $p && -f "$p/Makefile") {
		$lookup{$name} = $p;
		return $p ;
	    }
	}
    }
}

sub reset_installed {
    #opendir(DIR,"$PREFIX/.gar") ;
    #my @packages = grep(!/^\./,readdir(DIR));
    #closedir(DIR);
    #for my $p (map(find_package($_),@packages)) {
    #    mysystem("make -C $p uninstall");
    #}
    mysystem("rm -rv $PREFIX/.gar/* $PREFIX/*");
}

sub random_sort {
    my @x = @_;
    my %u; for my $i (@x) { $u{$i} = rand() } ;
    return sort {$u{$a} <=> $u{$b}} @x;
}

sub find_children {
    my $pid = shift;
    my @children = `pgrep -s 0 -P $pid`; 
    chomp(@children);
    print "pgrep -s 0 -P $pid: ", join(" ", @children), "\n";
    return @children;
}

sub find_children_recursive {
    my $pid = shift;
    my %seen;
    my @search = find_children($pid);
    my @children;
    while (my $c = pop @search) {
	next if $seen{$c} ;
	push(@children, $c);
	$seen{$c} = 1;
	my @extra = find_children($c);
	push(@search, @extra) if @extra;
    };
    return @children;
}
    
sub find_pkg {
    my $name = shift;
    my @dirs = ("$name", "gnu/$name", "deps/$name", glob("*/$name"));
    print "$name dirs = @dirs\n";
    my $pkg=(grep(-d $_ && -f "$_/Makefile",@dirs))[0];
    if ($pkg eq undef) {
	die "can't find <$pkg> for <*/$name>";
    };
    return $pkg;
}

use Digest::MD5;

sub compute_hash {
    my $name = shift;
    my @deps = @_;
    my $md5 = Digest::MD5->new;
    my $dir = find_pkg($name);
    for my $file ("$dir/Makefile","$dir/sha256sums", "$dir/md5sums", map(hashfile($_),@deps)) {
	next if ! -f $file;
	warn "hashing $file\n";
	open(FILE,"<$file") || die "can't open $file: $!";
	$md5->addfile(*FILE);
	close(FILE);
    };
    return $md5->hexdigest;
}

sub read_hash {
    my $file = shift;
    open (FILE, "<$file");
    my $stored_hash = <FILE>;
    chomp($stored_hash);
    close (FILE);
    return $stored_hash;
}

sub extract_options {
    my $conf;
    for (@_) {
	do { $conf =  $1; last ; } if /^cd.*?\/configure (.*)/;
    }
    $conf =~ s/--prefix=\S+\s*//;
    return $conf;
}
